src/java.base/share/classes/java/io/Console.java

```diff
@@ -592,25 +595,43 @@ public int read(char[] cbuf, int offset, int length)
 
         CHARSET = cs;
 
+        cons = instantiateConsole(istty);
+
         // Set up JavaIOAccess in SharedSecrets
         SharedSecrets.setJavaIOAccess(new JavaIOAccess() {
             public Console console() {
-                if (istty) {
-                    if (cons == null)
-                        cons = new Console();
-                    return cons;
-                }
-                return null;
-            }
-
-            public Charset charset() {
-                return CHARSET;
+                return cons;
             }
         });
     }
-    private static Console cons;
+
+    @SuppressWarnings("removal")
+    private static Console instantiateConsole(boolean istty) {
+        try {
+            // Try loading providers
+            PrivilegedAction<Console> pa = () -> {
+                var consModName = System.getProperty("jdk.console",
+                        JdkConsoleProvider.DEFAULT_PROVIDER_MODULE_NAME);
+                return ServiceLoader.load(JdkConsoleProvider.class).stream()
+                        .map(ServiceLoader.Provider::get)
+                        .filter(jcp -> consModName.equals(jcp.getClass().getModule().getName()))
+                        .map(jcp -> jcp.console(istty))
+                        .filter(Objects::nonNull)
+                        .findAny()
+                        .map(jc -> (Console) new ProxyingConsole(jc))
+                        .orElse(istty ? new Console() : null);
+            };
+            return AccessController.doPrivileged(pa);
+        } catch (ServiceConfigurationError ignore) {
```

Comment:
Should we perhaps just catch `Throwable` here since it's possible that the `PrivelegedAction` code could throw unchecked exception (for example the call to `JdkConsoleProvider.console()` could, in theory, lead to any kind of unchecked exceptions or errors like `NoClassDefFoundError`).